src/calling/haplotypes.rs

```diff
@@ -177,49 +184,93 @@ impl KallistoEstimates {
+            let gts = record.genotypes()?; //genotypes of all samples
 
-            for (sample_index, mut x) in sample.iter().zip(header.samples().into_iter()) {
-                for gta in gts.get(*sample_index).iter() {
+            let mut bv: BitVec<usize> = BitVec::new();
+            for (sample_index, x) in header.samples().into_iter().enumerate() {
+                for gta in gts.get(sample_index).iter() {
                     if gta == &Unphased(1) {
-                        let mut s = String::new();
-                        x.read_to_string(&mut s);
-                        haplotypes.push(Haplotype(s));
+                        bv.push(true);
                     } else {
+                        bv.push(false);
                         break;
                     }
                 }
             }
-            let mut haplotype_variants = HashMap::new();
-            haplotype_variants.insert(Variant(variant.clone()), haplotypes);
-            let haplotype_variants = HaplotypeVariants(haplotype_variants);
-            vec_haplotype_variants.push(haplotype_variants);
+            candidate_variants.insert(Variant(variant.clone()), bv);
+        }
+        Ok(HaplotypeVariants(candidate_variants))
+    }
+}
+
+#[derive(Derefable, Debug, Clone, PartialEq, Eq)]
+pub(crate) struct HaplotypeCalls(#[deref] HashMap<Variant, String>);
+
+impl HaplotypeCalls {
+    pub(crate) fn new(haplotype_calls: &mut bcf::Reader) -> Result<Self> {
+        let mut calls = HashMap::new();
+        for record_result in haplotype_calls.records() {
+            let record = record_result?;
+
+            //store the variant and the afd if the prob_present field is not nan (to save space)
+            if !record.info(b"PROB_PRESENT").float()?.unwrap()[0].is_nan() {
+                let mut variant = String::new();
+                variant.push_str(&format!("{}{}", record.contig(), "_"));
+                let pos = record.pos().to_string();
+                variant.push_str(&format!("{}{}", pos, "_"));
+                for allele in record.alleles() {
+                    if allele.len() == 1 {
+                        for c in allele {
+                            let base = char::from(*c).to_string();
+                            variant.push_str(&format!("{}{}", base, "_"));
+                        }
+                    } else {
+                        //the case for structural variants
+                        for c in allele {
+                            let base = char::from(*c).to_string();
+                            variant.push_str(&base);
+                        }
+                        variant.push_str("_")
+                    }
+                }
+
+                //store the afd string
+                let afd_utf = record.format(b"AFD").string()?[0];
+                let afd = std::str::from_utf8(afd_utf).unwrap();
+                calls.insert(Variant(variant.clone()), afd.to_string());
```

Comment:
I don't think we should keep the string here. Rather store this as a `HashMap<VariantId, HashMap<AlleleFreq, LogProb>>` here (with VariantId being the ID you get from record.id()). This way, you avoid recomputing this in the model methods, where it would be unnecessarily expensive because they are evaluated multiple times.